File: ios/gigyareactnative/GigyaBridge.m
//
//  GigyaBridge.m
//  gigyareactnative
//
//  Created by Gheerish Bansooodeb on 7/25/17.
//  Copyright © 2017 Facebook. All rights reserved.
//

#import "GigyaBridge.h"
#import "AppDelegate.h"
#import <GigyaSDK/Gigya.h>
#import "UIViewController+Utils.h"

@implementation GigyaBridge
GSPluginView *pluginView, *commentView;
bool pluginShowing;

RCT_EXPORT_MODULE();

// Gigya React Native custom events
- (NSArray<NSString *> *)supportedEvents
{
  return @[@"AccountDidLogin", @"AccountDidLogout",@"PluginViewFiredEvent",@"PluginViewDidFailWithError"];
}

- (dispatch_queue_t)methodQueue
{
  return dispatch_get_main_queue();
}

// Gigya Accounts Delegates
- (void)accountDidLogin:(GSAccount *)account{
  [self sendEventWithName:@"AccountDidLogin" body:account.JSONString];
  if(pluginView != nil){
    [pluginView removeFromSuperview];
  }
}

- (void)accountDidLogout{
  [self sendEventWithName:@"AccountDidLogout" body:@"AccountDidLogout"];
}

// Gigya React Native Bridge methods
RCT_EXPORT_METHOD(initBridge) {
  [Gigya setAccountsDelegate:self];
}

RCT_EXPORT_METHOD(login:(NSString *)loginId password:(NSString *)password callback:(RCTResponseSenderBlock)callback) {
  NSMutableDictionary *userAction = [NSMutableDictionary dictionary];
  [userAction setObject:loginId forKey:@"loginID"];
  [userAction setObject:password forKey:@"password"];

  GSRequest *request = [GSRequest requestForMethod:@"accounts.login" parameters:userAction];
  [request sendWithResponseHandler:^(GSResponse *response, NSError *error) {
    if (error) {
      callback(@[error.localizedDescription, [NSNull null]]);
    }
  }];
}

RCT_EXPORT_METHOD(socialLogin:(NSString *)provider callback:(RCTResponseSenderBlock)callback) {
   NSMutableDictionary *parameters = [NSMutableDictionary dictionary];
  [parameters setObject:@"saveProfileAndFail" forKey:@"x_conflictHandling"];

  [Gigya loginToProvider:provider
              parameters:parameters
                    over:[UIViewController currentViewController]
       completionHandler:^(GSUser *user, NSError *error) {

         if (error) {
           callback(@[error.localizedDescription, [NSNull null]]);
         }
       }
   ];
}

RCT_EXPORT_METHOD(isSessionValid:(RCTResponseSenderBlock)callback) {
  BOOL isValid = [Gigya isSessionValid];
  callback(@[[NSNull null], @(isValid)]);
}

RCT_EXPORT_METHOD(getAccountInfo:(RCTResponseSenderBlock)callback) {
  GSRequest *request = [GSRequest requestForMethod:@"accounts.getAccountInfo"];
  
  [request sendWithResponseHandler:^(GSResponse *response, NSError *error) {
    if (!error) {
      callback(@[[NSNull null], response.JSONString]);
    } else {
      callback(@[error.localizedDescription, [NSNull null]]);
    }
  }];
}

RCT_EXPORT_METHOD(setAccountInfo:(NSDictionary*)params callback:(RCTResponseSenderBlock)callback) {
  GSRequest *request = [GSRequest requestForMethod:@"accounts.setAccountInfo" parameters:params];
  [request sendWithResponseHandler:^(GSResponse *response, NSError *error) {
    if (!error) {
      callback(@[[NSNull null], response.JSONString]);
    } else {
      callback(@[error.localizedDescription, [NSNull null]]);
    }
  }];
}

RCT_EXPORT_METHOD(logout) {
  [Gigya logout];
}

RCT_EXPORT_METHOD(resetPassword:(NSString *)emailAddress callback:(RCTResponseSenderBlock)callback) {
  NSMutableDictionary *userAction = [NSMutableDictionary dictionary];
  [userAction setObject:emailAddress forKey:@"loginID"];
  [userAction setObject:emailAddress forKey:@"email"];
  
  GSRequest *request = [GSRequest requestForMethod:@"accounts.resetPassword" parameters:userAction];
  [request sendWithResponseHandler:^(GSResponse *response, NSError *error) {
    if (!error) {
      callback(@[[NSNull null], response.JSONString]);
    } else {
      callback(@[error.localizedDescription, [NSNull null]]);
    }
  }];
}

RCT_EXPORT_METHOD(registerFlow:(NSString *)firstName lastName:(NSString *)lastName email:(NSString *)email password:(NSString *)password tnc:(NSString *)tnc callback:(RCTResponseSenderBlock)callback) {
  GSRequest *request = [GSRequest requestForMethod:@"accounts.initRegistration" parameters:nil];
  [request sendWithResponseHandler:^(GSResponse *response, NSError *error) {
    if (!error) {
      NSString *regToken = response[@"regToken"];

      NSMutableDictionary *userAction = [NSMutableDictionary dictionary];
      [userAction setObject:email forKey:@"email"];
      [userAction setObject:password forKey:@"password"];
      [userAction setObject:regToken forKey:@"regToken"];
      
      
      GSRequest *request = [GSRequest requestForMethod:@"accounts.register" parameters:userAction];
      [request sendWithResponseHandler:^(GSResponse *response, NSError *error) {
        if (error.code == GSErrorAccountPendingRegistration) {
          
          //Get new regToken from response as an error was thrown
          NSString *regToken2 = response[@"regToken"];
          
          NSString *profileData = [NSString stringWithFormat: @"{ 'firstName' : '%@' , 'lastName' : '%@' }", firstName, lastName];
          NSString *dataData = [NSString stringWithFormat: @"{ 'terms' : '%@' }", tnc] ;
          
          NSMutableDictionary *userAction2 = [NSMutableDictionary dictionary];
          [userAction2 setObject:profileData forKey:@"profile"];
          [userAction2 setObject:dataData forKey:@"data"];
          [userAction2 setObject:regToken2 forKey:@"regToken"];
          
          GSRequest *request = [GSRequest requestForMethod:@"accounts.setAccountInfo" parameters:userAction2];
          [request sendWithResponseHandler:^(GSResponse *response, NSError *error) {
            if (!error) {
              
              NSMutableDictionary *userAction3 = [NSMutableDictionary dictionary];
              [userAction3 setObject:regToken2 forKey:@"regToken"];
              
              GSRequest *request = [GSRequest requestForMethod:@"accounts.finalizeRegistration" parameters:userAction3];
              [request sendWithResponseHandler:^(GSResponse *response, NSError *error) {
                if (!error) {
                  
                  //Login the user
                  NSMutableDictionary *userAction4 = [NSMutableDictionary dictionary];
                  [userAction4 setObject:email forKey:@"loginID"];
                  [userAction4 setObject:password forKey:@"password"];
                  
                  GSRequest *request = [GSRequest requestForMethod:@"accounts.login" parameters:userAction4];
                  [request sendWithResponseHandler:^(GSResponse *response, NSError *error) {
                    if (error) {
                      callback(@[error.localizedDescription, [NSNull null]]);
                    }
                  }];
                }
                else {
                  callback(@[error.localizedDescription, [NSNull null]]);
                }
              }];
            }
            else {
              callback(@[error.localizedDescription, [NSNull null]]);
            }
          }];
        }
        else {
          callback(@[error.localizedDescription, [NSNull null]]);
        }
      }];
    }
    else {
      callback(@[error.localizedDescription, [NSNull null]]);
    }
  }];
}

RCT_EXPORT_METHOD(showComments:(NSString *)categoryID streamID:(NSString *)streamID dimensions:(NSDictionary *)dimensions callback:(RCTResponseSenderBlock)callback) {
    
    //Parse params
    NSNumber* x = [dimensions objectForKey:@"x"];
    NSNumber* y = [dimensions objectForKey:@"y"];
    NSNumber* w = [dimensions objectForKey:@"w"];
    NSNumber* h = [dimensions objectForKey:@"h"];

    UIView *currentView = [[[[UIApplication sharedApplication] keyWindow] subviews] lastObject].superview;
    CGRect region = CGRectMake([x intValue],[y intValue],[w intValue],[h intValue]);
  
    NSInteger version = 2;
    NSMutableDictionary *pluginParams = [NSMutableDictionary dictionary];
    [pluginParams setObject:categoryID forKey:@"categoryID"];
    [pluginParams setObject:streamID forKey:@"streamID"];
    [pluginParams setObject:[NSNumber numberWithInteger:version] forKey:@"version"];
    pluginView = [[GSPluginView alloc] initWithFrame:region];
    pluginView.delegate = self;
    [pluginView loadPlugin:@"comments.commentsUI" parameters:pluginParams];
  
    [currentView addSubview:pluginView];
    pluginShowing = YES;
    callback(@[[NSNull null], @"Plugin loaded"]);
}

RCT_EXPORT_METHOD(hidePluginView) {
  if(pluginShowing){
    [pluginView removeFromSuperview];
  }
}

//Gigya Plugin View support
RCT_EXPORT_METHOD(showScreenSet:(NSString *)screensetName parameters:(NSDictionary *)params callback:(RCTResponseSenderBlock)callback) {
  
  //Parse params
  NSNumber* x = [params objectForKey:@"x"];
  NSNumber* y = [params objectForKey:@"y"];
  NSNumber* w = [params objectForKey:@"w"];
  NSNumber* h = [params objectForKey:@"h"];
  
  NSDictionary* screenSetDictionaryFromReact = [params objectForKey:@"screenSetParams"];
  NSMutableDictionary *screenSetParams = [screenSetDictionaryFromReact mutableCopy];
  [screenSetParams setObject:screensetName forKey:@"screenSet"];

  UIView *currentView = [[[[UIApplication sharedApplication] keyWindow] subviews] lastObject].superview;
  CGRect region = CGRectMake([x intValue],[y intValue],[w intValue],[h intValue]);
  
  pluginView = [[GSPluginView alloc] initWithFrame:region];
  pluginView.delegate = self;
  pluginView.showLoginProgress = YES;
  [pluginView loadPlugin:@"accounts.screenSet" parameters:screenSetParams];
  
  [currentView addSubview:pluginView];
  
//  [Gigya showPluginDialogOver:[UIViewController currentViewController] plugin:@"accounts.screenSet" parameters:params
//            completionHandler:nil delegate:self];
  callback(@[[NSNull null], @"Plugin loaded"]);
}

//Plugin Delegates
- (void)pluginView:(GSPluginView *)pluginView2 firedEvent:(NSDictionary *)event
{
    NSLog(@"Plugin event from %@ - %@", pluginView2.plugin, [event objectForKey:@"eventName"]);
  
    //detect afterSubmit event of screenset - and close screen natively
    if([pluginView2.plugin  isEqual: @"accounts.screenSet"] &&  [[event objectForKey:@"eventName"]  isEqual: @"afterSubmit"]){
      [pluginView removeFromSuperview];
      pluginView = nil;
    }
  
    NSError *error;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:event
                                                     options:NSJSONWritingPrettyPrinted
                                                       error:&error];
    if (! jsonData) {
      NSLog(@"Got an error: %@", error);
      [self sendEventWithName:@"PluginViewFiredEvent" body:error];
    } else {
      NSString *jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
      [self sendEventWithName:@"PluginViewFiredEvent" body:jsonString];
    }
}

- (void)pluginView:(GSPluginView *)pluginView finishedLoadingPluginWithEvent:(NSDictionary *)event
{
  NSLog(@"Finished loading plugin: %@", pluginView.plugin);
}

- (void)pluginView:(GSPluginView *)pluginView didFailWithError:(NSError *)error
{
  NSLog(@"Plugin error: %@", [error localizedDescription]);
  [self sendEventWithName:@"PluginViewFiredEvent" body:[error localizedDescription]];
  [pluginView removeFromSuperview];
}


@end

File: android/app/src/main/java/com/gigyareactnative/GigyaBridge.java
package com.gigyareactnative;

import android.support.v4.app.FragmentTransaction;
import android.util.Log;
import android.support.annotation.Nullable;

import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.ReadableMapKeySetIterator;
import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Callback;
import com.facebook.react.module.annotations.ReactModule;

import com.gigya.socialize.GSKeyNotFoundException;
import com.gigya.socialize.GSObject;
import com.gigya.socialize.GSResponse;
import com.gigya.socialize.GSResponseListener;
import com.gigya.socialize.android.GSAPI;
import com.gigya.socialize.android.event.GSAccountsEventListener;
import com.gigya.socialize.android.GSPluginFragment;
import com.gigya.socialize.android.event.GSPluginListener;

import org.json.JSONException;
import org.json.JSONObject;
import java.util.Map;

/**
 * Created by gheerish bansoodeb on 16/08/2017.
 */

public class GigyaBridge extends ReactContextBaseJavaModule {

    ReactApplicationContext thisReactContext;

    public GigyaBridge(ReactApplicationContext reactContext) {
        super(reactContext);
        thisReactContext = reactContext;
    }

    private void sendEvent(ReactContext reactContext,
                           String eventName,
                           @Nullable String params) {
        reactContext
                .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                .emit(eventName, params);
    }

    @Override
    public String getName() {
        return "GigyaBridge";
    }

    @ReactMethod
    public void initBridge() {
        GSAPI.getInstance().setAccountsEventListener(new GSAccountsEventListener() {
            @Override
            public void onLogin(GSObject user, Object context) {
                sendEvent(thisReactContext, "AccountDidLogin", user.toJsonString());
            }

            @Override
            public void onLogout(Object context) {
                sendEvent(thisReactContext, "AccountDidLogout", "AccountDidLogout");
            }
        });
    }

    @ReactMethod
    public void isSessionValid(Callback callback) {
        boolean isValid = false;

        if(GSAPI.getInstance().getSession() != null){
            isValid = GSAPI.getInstance().getSession().isValid();
        }
        callback.invoke(null, isValid);
    }

    @ReactMethod
    public void login(String loginId, String password, final Callback callback){
        GSObject params = new GSObject();
        params.put("loginID", loginId);
        params.put("password",password);

        GSResponseListener resListener = new GSResponseListener() {
            @Override
            public void onGSResponse(String method, GSResponse response, Object context) {
                try {
                    if (response.getErrorCode()!=0) {
                        callback.invoke(response.getErrorMessage());
                    }
                }
                catch (Exception ex) {
                    callback.invoke(ex.getLocalizedMessage());
                }
            }
        };

        String methodName = "accounts.login";
        GSAPI.getInstance().sendRequest(methodName, params, resListener, null);
    }

    @ReactMethod
    public void resetPassword(String emailAddress, final Callback callback) {
        GSObject params = new GSObject();
        params.put("loginID", emailAddress);
        params.put("email",emailAddress);

        GSResponseListener resListener = new GSResponseListener() {
            @Override
            public void onGSResponse(String method, GSResponse response, Object context) {
                try {
                    if (response.getErrorCode()!=0) {
                        callback.invoke(response.getErrorMessage());
                    } else {
                        callback.invoke(null, response.getData().toJsonString());
                    }
                }
                catch (Exception ex) {
                    callback.invoke(ex.getLocalizedMessage());
                }
            }
        };

        String methodName = "accounts.resetPassword";
        GSAPI.getInstance().sendRequest(methodName, params, resListener, null);
    }

    @ReactMethod
    public void registerFlow(final String firstName, final String lastName, final String email, final String password, final String tnc, final Callback callback) {
        GSResponseListener resListener = new GSResponseListener() {
            @Override
            public void onGSResponse(String method, GSResponse response, Object context) {
                try {
                    if (response.getErrorCode()==0) {
                        String regToken = response.getData().getString("regToken");

                        // Step 1 - Defining request parameters
                        GSObject params = new GSObject();
                        params.put("regToken", regToken);
                        params.put("email", email);
                        params.put("password", password);
                        params.put("finalizeRegistration", true);

                        GSObject profile = new GSObject();
                        profile.put("firstName", firstName);
                        profile.put("lastName", lastName);
                        params.put("profile", profile);

                        GSObject data = new GSObject();
                        data.put("terms", tnc);
                        params.put("data", data);

                        // Step 2 - Defining response listener. The response listener will handle the request's response.
                        GSResponseListener resListener = new GSResponseListener() {
                            @Override
                            public void onGSResponse(String method, GSResponse response, Object context) {
                                try {
                                    if (response.getErrorCode()==0) {
                                        callback.invoke(null, response.getData().toJsonString());
                                    } else {
                                        callback.invoke(response.getErrorMessage());
                                    }
                                }
                                catch (Exception ex) {  ex.printStackTrace();  }
                            }
                        };

                        // Step 3 - Sending the request
                        String methodName = "accounts.register";
                        GSAPI.getInstance().sendRequest(methodName, params, resListener, null);

                    } else {
                        callback.invoke(response.getErrorMessage());
                    }
                }
                catch (Exception ex) {  ex.printStackTrace();  }
            }
        };

        // Step 3 - Sending the request
        String methodName = "accounts.initRegistration";
        GSAPI.getInstance().sendRequest(methodName, null, resListener, null);
    }

    @ReactMethod
    public void getAccountInfo(final Callback callback){
        GSResponseListener resListener = new GSResponseListener() {
            @Override
            public void onGSResponse(String method, GSResponse response, Object context) {
                try {
                    if (response.getErrorCode()!=0) {
                        callback.invoke(response.getData().toJsonString());
                    } else {
                        callback.invoke(null, response.getData().toJsonString());
                    }
                }
                catch (Exception ex) {
                    callback.invoke(ex.getLocalizedMessage());
                }
            }
        };

        String methodName = "accounts.getAccountInfo";
        GSAPI.getInstance().sendRequest(methodName, null, resListener, null);
    }

    @ReactMethod
    public void socialLogin(String provider, final Callback callback){
        GSObject params = new GSObject();
        params.put("provider", provider);

        GSResponseListener resListener = new GSResponseListener() {
            @Override
            public void onGSResponse(String method, GSResponse response, Object context) {
                try {
                    if (response.getErrorCode()!=0) {
                        callback.invoke(response.getData().toJsonString());
                    }
                }
                catch (Exception ex) {
                    callback.invoke(ex.getLocalizedMessage());
                }
            }
        };

        try {
        GSAPI.getInstance().login(thisReactContext.getCurrentActivity(), params, resListener , false, null);
        }
        catch (Exception ex) {
            callback.invoke(ex.getLocalizedMessage());
        }
    }

    @ReactMethod
    public void logout(){
        GSAPI.getInstance().logout();
    }

    @ReactMethod
    public void showScreenSet(String screensetName, ReadableMap screensetParams, final Callback callback){
        GSObject params = new GSObject();

        Map<String,Object> reactMap = JsonConvert.reactToMap(screensetParams);
        Map<String,Object> screensetMap = (Map<String,Object>)reactMap.get("screenSetParams");

        for (Map.Entry<String, Object> entry : screensetMap.entrySet())
        {
            params.put("entry.getKey()", entry.getValue());
        }
        params.put("screenSet", screensetName);

        GSAPI.getInstance().showPluginDialog("accounts.screenSet", params, new GSPluginListener() {
            @Override
            public void onLoad(GSPluginFragment gsPluginFragment, GSObject gsObject) {
                callback.invoke(null, "Plugin loaded");
            }

            @Override
            public void onError(GSPluginFragment gsPluginFragment, GSObject error) {
                callback.invoke(error.toJsonString());
            }

            @Override
            public void onEvent(GSPluginFragment gsPluginFragment, GSObject gsObject) {
            }
        }, null);
    }

    @ReactMethod
    public void showComments(String categoryID, String streamID, ReadableMap dimensions, final Callback callback){
        GSObject params = new GSObject();
        params.put("categoryID", categoryID);
        params.put("streamID", streamID);
        params.put("version", 2);

        GSAPI.getInstance().showPluginDialog("comments.commentsUI", params, new GSPluginListener() {
            @Override
            public void onLoad(GSPluginFragment gsPluginFragment, GSObject gsObject) {
                callback.invoke(null, "Plugin loaded");
            }

            @Override
            public void onError(GSPluginFragment gsPluginFragment, GSObject error) {
                callback.invoke(error.toJsonString());
            }

            @Override
            public void onEvent(GSPluginFragment gsPluginFragment, GSObject gsObject) {
            }
        }, null);
    }

    @ReactMethod
    public void hidePluginView(){
    }
}